src/sql/src/plan/lowering.rs

```diff
@@ -1148,6 +1285,39 @@ where
             });
         }
     });
+    // Collect all the outer columns referenced by any CTE referenced by
+    // the inner relation.
+    inner.visit(&mut |e| match e {
```

Comment:
We could do a single traversal if we split `visit_columns` into `visit_with_depth_tracking` and `visit_columns_in_relation` (non-recursive).